### PR TITLE
Evaluate keccak256 of string literals at compile-time.

### LINKED ISSUE #4024: Compile time evaluation of constant hashes
There's room for improvement in optimising compile-time evaluation of constant hashes when optimisation is enabled.

```
contract Hash{

    bytes32 constant HASH1 = keccak256('');
    bytes32 constant HASH2 = keccak256('12345678901234567890123456789012');
    bytes32 constant HASH3 = keccak256(uint256(0));
    
    function foo1() public pure returns (bytes32) {
        return HASH1;
    }

    function foo2() public pure returns (bytes32) {
        return HASH2;
    }

    function foo3() public pure returns (bytes32) {
        return HASH3;
    }
}
```

With optimisation off, none of the constant hashes is evaluated at compile-time. In each case the `keccak256` operation is being done at run time. This is somewhat expected.

With optimisation enabled, only the first (`keccak256('')`) is evaluated at compile time. The remainder (`keccak256('12345678901234567890123456789012')`, `keccak256(uint256(0))`) are evaluated at run time, as if optimisation were off. This is unexpected. Expected behaviour would be to evaluate all constant hashes at compile time.

Test environment: Remix, Solidity version 0.4.23+commit.124ca40d
Conclusions reached by inspecting the generated byte code. It's quite obvious if the hash is present or not.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/4024

(draft in order not to disturb the 0.6.11 release process)

### CODE REVIEW

**Review by leonardoalt — APPROVED** [review]
Looks good!

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -1020,18 +1020,28 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 
 		ArrayType const* arrayType = TypeProvider::bytesMemory();
 
-		auto array = convert(*arguments[0], *arrayType);
+		if (auto const* stringLiteral = dynamic_cast<StringLiteralType const*>(arguments.front()->annotation().type))
+		{
+			// Optimization: Compute keccak256 on string literals at compile-time.
```
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1025]: Nicer would of course be if we had yul memory objects, would have the optimizer realize that the entire memory object has known contents and then have the optimizer first replace ``keccak256`` with its evaluation and then remove the memory object as unused :-)... but well...

**Review by ekpyron — APPROVED** [review]
And when do we generalize this to https://github.com/ethereum/solidity/issues/3878 ;-)?

**Review by axic — APPROVED** [review]

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1025]: Yes, that would be very nice indeed! :)

**Pull request merged**